library/src/amd_detail/hipsolver_refactor.cpp

```diff
@@ -0,0 +1,952 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IM-
+ * PLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNE-
+ * CTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ *
+ * ************************************************************************ */
+
+/*! \file
+ *  \brief Implementation of the compatibility refactor APIs that require especial
+ *  calls to hipSOLVER or rocSOLVER.
+ */
+
+#include "error_macros.hpp"
+#include "exceptions.hpp"
+#include "hipsolver.h"
+#include "hipsolver_conversions.hpp"
+
+#include "rocblas/internal/rocblas_device_malloc.hpp"
+#include "rocblas/rocblas.h"
+#include "rocsolver/rocsolver.h"
+#include <algorithm>
+#include <climits>
+#include <functional>
+#include <iostream>
+#include <math.h>
+
+extern "C" {
+
+/******************** HANDLE ********************/
+struct hipsolverRfHandle
+{
+    hipsolverRfResetValuesFastMode_t fast_mode;
+    hipsolverRfMatrixFormat_t        matrix_format;
+    hipsolverRfUnitDiagonal_t        diag_format;
+    hipsolverRfNumericBoostReport_t  numeric_boost;
+
+    hipsolverRfFactorization_t   fact_alg;
+    hipsolverRfTriangularSolve_t solve_alg;
+
+    rocblas_handle   handle;
+    rocsolver_rfinfo rfinfo;
+
+    rocblas_int n, nnzA, nnzL, nnzU, nnzLU, batch_count;
+    double      effective_zero;
+    double      boost_val;
+
+    rocblas_int* dPtrA;
+    rocblas_int* dIndA;
+    double*      dValA;
+
+    rocblas_int *dPtrL, *hPtrL;
+    rocblas_int *dIndL, *hIndL;
+    double *     dValL, *hValL;
+
+    rocblas_int *dPtrU, *hPtrU;
+    rocblas_int *dIndU, *hIndU;
+    double *     dValU, *hValU;
+
+    rocblas_int *dPtrLU, *hPtrLU;
+    rocblas_int *dIndLU, *hIndLU;
+    double *     dValLU, *hValLU;
+
+    rocblas_int* dP;
+    rocblas_int* dQ;
+
+    uint8_t *d_buffer, *h_buffer;
+
+    // Constructor
+    explicit hipsolverRfHandle()
+        : fast_mode(HIPSOLVERRF_RESET_VALUES_FAST_MODE_OFF)
+        , matrix_format(HIPSOLVERRF_MATRIX_FORMAT_CSR)
+        , diag_format(HIPSOLVERRF_UNIT_DIAGONAL_STORED_L)
+        , numeric_boost(HIPSOLVERRF_NUMERIC_BOOST_NOT_USED)
+        , fact_alg(HIPSOLVERRF_FACTORIZATION_ALG0)
+        , solve_alg(HIPSOLVERRF_TRIANGULAR_SOLVE_ALG1)
+        , n(0)
+        , nnzA(0)
+        , nnzL(0)
+        , nnzU(0)
+        , nnzLU(0)
+        , batch_count(0)
+        , effective_zero(0.0)
+        , boost_val(0.0)
+        , d_buffer(nullptr)
+        , h_buffer(nullptr)
+    {
+    }
+
+    // Allocate device memory
+    hipsolverStatus_t malloc_device(int n, int nnzA, int nnzL, int nnzU)
```

Comment:
This is a lot more gnarly than rocsolver. I don't see any obvious issues with this, but I definitely was wrong when I assumed that the resource management in hipSOLVER was similar to rocSOLVER.